src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
-#ifdef ENABLE_PARALLELIZATION
-        std::unique_lock<std::mutex> lock(link_list_locks_[selectedNeighbor]);
-#endif
-        idType *neighbor_neighbors = get_linklist_at_level(selectedNeighbor, level);
-        linkListSize sz_link_list_other = getListCount(neighbor_neighbors);
-
-        if (sz_link_list_other > Mcurmax)
-            throw std::runtime_error("Bad value of sz_link_list_other");
-        if (selectedNeighbor == cur_c)
-            throw std::runtime_error("Trying to connect an element to itself");
-        if (level > element_levels_[selectedNeighbor])
-            throw std::runtime_error("Trying to make a link on a non-existent level");
-
-        // If the selected neighbor can add another link (hasn't reached the max) - add it.
-        if (sz_link_list_other < Mcurmax) {
-            neighbor_neighbors[sz_link_list_other] = cur_c;
-            setListCount(neighbor_neighbors, sz_link_list_other + 1);
+template <typename DataType, typename DistType>
+idType HNSWIndex<DataType, DistType>::mutuallyConnectNewElement(
+    idType new_node_id, candidatesMaxHeap<DistType> &top_candidates, size_t level) {
+
+    // The maximum number of neighbors allowed for an existing neighbor (not new).
+    size_t max_M_cur = level ? maxM_ : maxM0_;
+
+    // Filter the top candidates to the selected neighbors by the algorithm heuristics.
+    getNeighborsByHeuristic2(top_candidates, M_);
+    assert(top_candidates.size() <= M_ &&
+           "Should be not be more than M_ candidates returned by the heuristic");
+
+    vecsim_stl::vector<idType> selected_neighbors(this->allocator);
+    selected_neighbors.reserve(M_);
+    while (!top_candidates.empty()) {
+        selected_neighbors.push_back(top_candidates.top().second);
+        top_candidates.pop();
+    }
+
+    // The closest vector that has found to be returned (and start the scan from it in the next
+    // level).
+    idType next_closest_entry_point = selected_neighbors.back();
+    idType *new_node_neighbors_list = get_linklist_at_level(new_node_id, level);
+    assert(getListCount(new_node_neighbors_list) == 0 &&
+           "The newly inserted element should have blank link list");
+
+    // Create the incoming edges for the new node in the current level.
+    auto *incoming_edges = new (this->allocator) vecsim_stl::vector<idType>(this->allocator);
+    setIncomingEdgesPtr(new_node_id, level, (void *)incoming_edges);
+
+    for (idType selected_neighbor : selected_neighbors) {
+        std::unique_lock<std::mutex> node_lock;
+        std::unique_lock<std::mutex> neighbor_lock;
+        idType lower_id = (new_node_id < selected_neighbor) ? new_node_id : selected_neighbor;
+        if (lower_id == new_node_id) {
+            node_lock = std::unique_lock<std::mutex>(element_neighbors_locks_[new_node_id]);
+            neighbor_lock =
+                std::unique_lock<std::mutex>(element_neighbors_locks_[selected_neighbor]);
         } else {
```

Comment:
This should be fenced somehow to prevent compiler or CPU reordering

Reply:
intersting. I'll look into that

Reply:
After checking with the experts, it's not required:) The compiler assures the fence between both locks implicitly